File: backend/seed.py
"""
Database seed script for Institute ERP.
Creates demo data matching the original Node.js seed (prisma/seed.ts).
"""
import uuid
from datetime import datetime, timedelta

from passlib.context import CryptContext
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

from app.database import Base
from app.models.announcement import Announcement
from app.models.attendance import Attendance
from app.models.batch import Batch
from app.models.batch_module import BatchModule
from app.models.counsellor import Counsellor
from app.models.counsellor_student import CounsellorStudent
from app.models.fee import Fee, FeePayment
from app.models.lecture import Lecture
from app.models.marks import Marks
from app.models.module import Module
from app.models.notification import Notification
from app.models.student import Student
from app.models.trainer import Trainer
from app.models.trainer_batch import TrainerBatch
from app.models.user import User

DATABASE_URL = "sqlite:///./dev.db"
engine = create_engine(DATABASE_URL, connect_args={"check_same_thread": False})
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

STUDENT_NAMES = [
    "Aarav Sharma", "Vivaan Gupta", "Aditya Singh", "Vihaan Patel",
    "Arjun Reddy", "Sai Kumar", "Reyansh Joshi", "Ayaan Khan",
    "Krishna Iyer", "Ishaan Nair", "Ananya Mishra", "Diya Chopra",
    "Myra Banerjee", "Sara Desai", "Aanya Mehta", "Aadhya Rao",
    "Navya Pillai", "Isha Saxena", "Riya Thakur", "Kavya Menon",
]


def generate_id():
    return str(uuid.uuid4())


def clear_database(db):
    """Delete all existing data in dependency order."""
    print("Clearing existing data...")
    db.query(FeePayment).delete()
    db.query(Fee).delete()
    db.query(Attendance).delete()
    db.query(Marks).delete()
    db.query(Notification).delete()
    db.query(Announcement).delete()
    db.query(CounsellorStudent).delete()
    db.query(Lecture).delete()
    db.query(BatchModule).delete()
    db.query(TrainerBatch).delete()
    db.query(Student).delete()
    db.query(Counsellor).delete()
    db.query(Trainer).delete()
    db.query(Module).delete()
    db.query(Batch).delete()
    db.query(User).delete()
    db.commit()


def seed():
    print("Creating tables...")
    Base.metadata.create_all(bind=engine)

    db = SessionLocal()
    try:
        clear_database(db)

        # --- Users ---
        print("Creating users...")

        # Admin
        admin_user = User(
            id=generate_id(),
            email="admin@institute.com",
            password=pwd_context.hash("admin123"),
            role="ADMIN",
            name="Admin User",
            phone="9000000001",
        )
        db.add(admin_user)

        # Trainers
        trainer_data = [
            {"email": "trainer1@institute.com", "name": "Rajesh Kumar", "specialization": "Full Stack Development"},
            {"email": "trainer2@institute.com", "name": "Priya Sharma", "specialization": "Frontend & UI/UX"},
            {"email": "trainer3@institute.com", "name": "Amit Patel", "specialization": "Backend & DevOps"},
        ]
        trainers = []
        for td in trainer_data:
            user = User(
                id=generate_id(),
                email=td["email"],
                password=pwd_context.hash("trainer123"),
                role="TRAINER",
                name=td["name"],
                phone=f"900000000{len(trainers) + 2}",
            )
            db.add(user)
            db.flush()
            trainer = Trainer(id=generate_id(), userId=user.id, specialization=td["specialization"])
            db.add(trainer)
            trainers.append(trainer)

        # Counsellors
        counsellor_data = [
            {"email": "counsellor1@institute.com", "name": "Sunita Reddy"},
            {"email": "counsellor2@institute.com", "name": "Deepak Verma"},
        ]
        counsellors = []
        for cd in counsellor_data:
            user = User(
                id=generate_id(),
                email=cd["email"],
                password=pwd_context.hash("counsellor123"),
                role="COUNSELLOR",
                name=cd["name"],
                phone=f"900000000{len(counsellors) + 5}",
            )
            db.add(user)
            db.flush()
            counsellor = Counsellor(id=generate_id(), userId=user.id)
            db.add(counsellor)
            counsellors.append(counsellor)

        db.flush()
        print(f"  Created 1 admin, {len(trainers)} trainers, {len(counsellors)} counsellors")

        # --- Batches ---
        print("Creating batches...")
        batch_data = [
            {"name": "Full Stack Web Dev - Batch 1", "startDate": datetime(2024, 1, 15), "endDate": datetime(2024, 7, 15)},
            {"name": "Full Stack Web Dev - Batch 2", "startDate": datetime(2024, 3, 1), "endDate": datetime(2024, 9, 1)},
            {"name": "Data Science - Batch 1", "startDate": datetime(2024, 2, 1), "endDate": datetime(2024, 8, 1)},
            {"name": "Mobile App Dev - Batch 1", "startDate": datetime(2024, 4, 1), "endDate": datetime(2024, 10, 1)},
        ]
        batches = []
        for bd in batch_data:
            batch = Batch(id=generate_id(), name=bd["name"], startDate=bd["startDate"], endDate=bd["endDate"], isActive=True)
            db.add(batch)
            batches.append(batch)
        db.flush()
        print(f"  Created {len(batches)} batches")

        # --- Modules ---
        print("Creating modules...")
        module_names = [
            "HTML & CSS",
            "JavaScript",
            "React.js",
            "Node.js & Express",
            "Database & SQL",
            "TypeScript",
            "DevOps & CI/CD",
            "Data Structures & Algorithms",
        ]
        modules = []
        for mn in module_names:
            module = Module(id=generate_id(), name=mn, description=f"Learn {mn}", duration=40)
            db.add(module)
            modules.append(module)
        db.flush()
        print(f"  Created {len(modules)} modules")

        # --- Batch-Module Assignments ---
        print("Creating batch-module assignments...")
        bm_count = 0
        for batch in batches:
            for module in modules:
                bm = BatchModule(id=generate_id(), batchId=batch.id, moduleId=module.id, status="IN_PROGRESS", completionPercent=25.0)
                db.add(bm)
                bm_count += 1
        db.flush()
        print(f"  Created {bm_count} batch-module assignments")

        # --- Trainer-Batch Assignments ---
        print("Creating trainer-batch assignments...")
        tb_assignments = [
            (trainers[0].id, batches[0].id),
            (trainers[0].id, batches[1].id),
            (trainers[1].id, batches[0].id),
            (trainers[1].id, batches[2].id),
            (trainers[2].id, batches[1].id),
            (trainers[2].id, batches[3].id),
        ]
        for trainer_id, batch_id in tb_assignments:
            tb = TrainerBatch(id=generate_id(), trainerId=trainer_id, batchId=batch_id)
            db.add(tb)
        db.flush()
        print(f"  Created {len(tb_assignments)} trainer-batch assignments")

        # --- Students ---
        print("Creating students...")
        students = []
        for i in range(20):
            user = User(
                id=generate_id(),
                email=f"student{i + 1}@institute.com",
                password=pwd_context.hash("student123"),
                role="STUDENT",
                name=STUDENT_NAMES[i],
                phone=f"98000000{i + 1:02d}",
            )
            db.add(user)
            db.flush()

            batch_index = i % 4
            counsellor_index = i % 2
            student = Student(
                id=generate_id(),
                userId=user.id,
                batchId=batches[batch_index].id,
                counsellorId=counsellors[counsellor_index].id,
                mode="OFFLINE" if i % 3 != 0 else "ONLINE",
            )
            db.add(student)
            students.append(student)
        db.flush()
        print(f"  Created {len(students)} students")

        # --- Counsellor-Student Assignments ---
        print("Creating counsellor-student assignments...")
        for i, student in enumerate(students):
            counsellor_index = i % 2
            cs = CounsellorStudent(
                id=generate_id(),
                counsellorId=counsellors[counsellor_index].id,
                studentId=student.id,
            )
            db.add(cs)
        db.flush()
        print(f"  Created {len(students)} counsellor-student assignments")

        # --- Lectures ---
        print("Creating lectures...")
        lectures = []
        base_date = datetime(2024, 2, 1)
        topics_list = [
            "Introduction to HTML",
            "CSS Box Model",
            "JavaScript Basics",
            "DOM Manipulation",
            "React Components",
            "State Management",
            "Node.js Fundamentals",
            "Express Routing",
            "SQL Queries",
            "Database Design",
            "TypeScript Types",
            "Interfaces and Generics",
            "Docker Basics",
            "CI/CD Pipelines",
            "Arrays and Strings",
            "Linked Lists",
            "Trees and Graphs",
            "Sorting Algorithms",
            "REST API Design",
            "Authentication & Authorization",
        ]
        for i in range(20):
            lecture = Lecture(
                id=generate_id(),
                batchId=batches[0].id,
                moduleId=modules[i % len(modules)].id,
                trainerId=trainers[0].id,
                date=base_date + timedelta(days=i),
                startTime="10:00",
                endTime="12:00",
                duration=120,
                topics=topics_list[i],
            )
            db.add(lecture)
            lectures.append(lecture)
        db.flush()
        print(f"  Created {len(lectures)} lectures")

        # --- Attendance ---
        print("Creating attendance records...")
        att_count = 0
        batch1_students = [s for s in students if s.batchId == batches[0].id]
        for lecture in lectures:
            for student in batch1_students:
                import random
                status = random.choice(["PRESENT", "PRESENT", "PRESENT", "ABSENT", "LATE"])
                att = Attendance(
                    id=generate_id(),
                    studentId=student.id,
                    lectureId=lecture.id,
                    status=status,
                    method="QR" if status == "PRESENT" else "MANUAL",
                )
                db.add(att)
                att_count += 1
        db.flush()
        print(f"  Created {att_count} attendance records")

        # --- Marks ---
        print("Creating marks...")
        marks_count = 0
        mark_types = ["QUIZ", "ASSIGNMENT", "PROJECT", "EXAM"]
        for student in students[:10]:
            for j, module in enumerate(modules[:4]):
                marks = Marks(
                    id=generate_id(),
                    studentId=student.id,
                    moduleId=module.id,
                    type=mark_types[j % len(mark_types)],
                    score=float(60 + (hash(student.id + module.id) % 35)),
                    maxScore=100.0,
                    remarks="Good performance" if (hash(student.id) % 2 == 0) else None,
                )
                db.add(marks)
                marks_count += 1
        db.flush()
        print(f"  Created {marks_count} marks records")

        # --- Fees ---
        print("Creating fees...")
        fee_count = 0
        for i, student in enumerate(students):
            total = 50000.0
            paid_amounts = [50000, 40000, 30000, 25000, 20000, 15000, 10000, 5000, 45000, 35000]
            paid = float(paid_amounts[i % len(paid_amounts)])
            pending = total - paid
            status = "PAID" if paid >= total else ("PARTIAL" if paid > 0 else "PENDING")

            fee = Fee(
                id=generate_id(),
                studentId=student.id,
                totalAmount=total,
                paidAmount=paid,
                pendingAmount=pending,
                dueDate=datetime(2024, 6, 30),
                status=status,
            )
            db.add(fee)
            db.flush()

            if paid > 0:
                payment = FeePayment(
                    id=generate_id(),
                    feeId=fee.id,
                    amount=paid,
                    paidAt=datetime(2024, 1, 15) + timedelta(days=i * 3),
                    method="UPI" if i % 3 == 0 else ("BANK_TRANSFER" if i % 3 == 1 else "CASH"),
                )
                db.add(payment)

            fee_count += 1
        db.flush()
        print(f"  Created {fee_count} fee records")

        # --- Notifications ---
        print("Creating notifications...")
        notif_count = 0
        for i, student in enumerate(students[:10]):
            notif = Notification(
                id=generate_id(),
                userId=student.userId,
                title="Welcome to Institute ERP",
                message=f"Welcome {STUDENT_NAMES[i]}! Your enrollment is confirmed.",
                type="INFO",
                isRead=i % 2 == 0,
            )
            db.add(notif)
            notif_count += 1

        # Notification for trainers
        for trainer in trainers:
            notif = Notification(
                id=generate_id(),
                userId=trainer.userId,
                title="New Batch Assigned",
                message="You have been assigned to a new batch. Check your dashboard for details.",
                type="INFO",
                isRead=False,
            )
            db.add(notif)
            notif_count += 1
        db.flush()
        print(f"  Created {notif_count} notifications")

        # --- Announcements ---
        print("Creating announcements...")
        announcements_data = [
            {"title": "Welcome to the New Semester", "message": "Classes begin next week. Please check your batch schedule.", "targetRole": "STUDENT"},
            {"title": "Faculty Meeting", "message": "All trainers are requested to attend the faculty meeting on Friday.", "targetRole": "TRAINER"},
            {"title": "Fee Payment Reminder", "message": "Please clear any pending fee payments before the end of this month.", "targetRole": "STUDENT"},
        ]
        for ad in announcements_data:
            ann = Announcement(
                id=generate_id(),
                createdById=admin_user.id,
                title=ad["title"],
                message=ad["message"],
                targetRole=ad["targetRole"],
                batchId=batches[0].id,
            )
            db.add(ann)
        db.flush()
        print(f"  Created {len(announcements_data)} announcements")

        db.commit()
        print("\nSeed completed successfully!")
        print("\n--- Default Credentials ---")
        print("Admin:      admin@institute.com / admin123")
        print("Trainer 1:  trainer1@institute.com / trainer123")
        print("Trainer 2:  trainer2@institute.com / trainer123")
        print("Trainer 3:  trainer3@institute.com / trainer123")
        print("Counsellor: counsellor1@institute.com / counsellor123")
        print("Counsellor: counsellor2@institute.com / counsellor123")
        print("Students:   student1@institute.com - student20@institute.com / student123")
        print("----------------------------")

    except Exception as e:
        db.rollback()
        print(f"Error during seeding: {e}")
        raise
    finally:
        db.close()


if __name__ == "__main__":
    seed()

File: backend/prisma/seed.ts
import { PrismaClient } from '@prisma/client';
import bcrypt from 'bcryptjs';

const prisma = new PrismaClient();

async function main() {
  console.log('Seeding database...');

  const password = await bcrypt.hash('admin123', 10);
  const trainerPassword = await bcrypt.hash('trainer123', 10);
  const counsellorPassword = await bcrypt.hash('counsellor123', 10);
  const studentPassword = await bcrypt.hash('student123', 10);

  // Create admin user
  const admin = await prisma.user.upsert({
    where: { email: 'admin@institute.com' },
    update: {},
    create: {
      email: 'admin@institute.com',
      password: password,
      name: 'Admin User',
      role: 'ADMIN',
      phone: '9000000001',
    },
  });
  console.log('Created admin:', admin.email);

  // Create 3 trainers
  const trainerData = [
    { email: 'trainer1@institute.com', name: 'Rajesh Kumar', specialization: 'Full Stack Development' },
    { email: 'trainer2@institute.com', name: 'Priya Sharma', specialization: 'Frontend & UI/UX' },
    { email: 'trainer3@institute.com', name: 'Amit Patel', specialization: 'Backend & DevOps' },
  ];

  const trainers = [];
  for (let i = 0; i < trainerData.length; i++) {
    const td = trainerData[i];
    const trainerUser = await prisma.user.upsert({
      where: { email: td.email },
      update: {},
      create: {
        email: td.email,
        password: trainerPassword,
        name: td.name,
        role: 'TRAINER',
        phone: `900000010${i + 1}`,
      },
    });
    const trainer = await prisma.trainer.upsert({
      where: { userId: trainerUser.id },
      update: {},
      create: {
        userId: trainerUser.id,
        specialization: td.specialization,
      },
    });
    trainers.push(trainer);
  }
  console.log('Created 3 trainers');

  // Create 2 counsellors
  const counsellorData = [
    { email: 'counsellor1@institute.com', name: 'Sunita Reddy' },
    { email: 'counsellor2@institute.com', name: 'Deepak Verma' },
  ];

  const counsellors = [];
  for (let i = 0; i < counsellorData.length; i++) {
    const cd = counsellorData[i];
    const counsellorUser = await prisma.user.upsert({
      where: { email: cd.email },
      update: {},
      create: {
        email: cd.email,
        password: counsellorPassword,
        name: cd.name,
        role: 'COUNSELLOR',
        phone: `900000020${i + 1}`,
      },
    });
    const counsellor = await prisma.counsellor.upsert({
      where: { userId: counsellorUser.id },
      update: {},
      create: {
        userId: counsellorUser.id,
      },
    });
    counsellors.push(counsellor);
  }
  console.log('Created 2 counsellors');

  // Create 4 batches
  const batchData = [
    { id: 'batch-fswd-2024-01', name: 'Full Stack Web Dev - Batch 1', startDate: new Date('2024-01-15'), endDate: new Date('2024-07-15') },
    { id: 'batch-fswd-2024-02', name: 'Full Stack Web Dev - Batch 2', startDate: new Date('2024-03-01'), endDate: new Date('2024-09-01') },
    { id: 'batch-data-2024-01', name: 'Data Science - Batch 1', startDate: new Date('2024-02-01'), endDate: new Date('2024-08-01') },
    { id: 'batch-mobile-2024-01', name: 'Mobile App Dev - Batch 1', startDate: new Date('2024-04-01'), endDate: new Date('2024-10-01') },
  ];

  const batches = [];
  for (const bd of batchData) {
    const batch = await prisma.batch.upsert({
      where: { id: bd.id },
      update: {},
      create: {
        id: bd.id,
        name: bd.name,
        startDate: bd.startDate,
        endDate: bd.endDate,
        isActive: true,
      },
    });
    batches.push(batch);
  }
  console.log('Created 4 batches');

  // Create 8 modules
  const moduleData = [
    { id: 'module-html-css', name: 'HTML & CSS', description: 'Fundamentals of web markup and styling', duration: 30 },
    { id: 'module-javascript', name: 'JavaScript', description: 'Core JavaScript programming concepts', duration: 45 },
    { id: 'module-react', name: 'React.js', description: 'Building modern UIs with React', duration: 40 },
    { id: 'module-nodejs', name: 'Node.js & Express', description: 'Server-side JavaScript with Node.js', duration: 35 },
    { id: 'module-database', name: 'Database & SQL', description: 'Relational databases and query optimization', duration: 30 },
    { id: 'module-typescript', name: 'TypeScript', description: 'Type-safe JavaScript development', duration: 25 },
    { id: 'module-devops', name: 'DevOps & CI/CD', description: 'Deployment pipelines and containerization', duration: 20 },
    { id: 'module-dsa', name: 'Data Structures & Algorithms', description: 'Problem solving and algorithm design', duration: 50 },
  ];

  const modules = [];
  for (const md of moduleData) {
    const mod = await prisma.module.upsert({
      where: { id: md.id },
      update: {},
      create: {
        id: md.id,
        name: md.name,
        description: md.description,
        duration: md.duration,
      },
    });
    modules.push(mod);
  }
  console.log('Created 8 modules');

  // Batch-module assignments (first batch gets all modules, others get subsets)
  const batchModuleAssignments = [
    { batchId: batches[0].id, moduleIds: modules.slice(0, 6).map(m => m.id), status: 'IN_PROGRESS', completion: 60 },
    { batchId: batches[1].id, moduleIds: modules.slice(0, 4).map(m => m.id), status: 'IN_PROGRESS', completion: 30 },
    { batchId: batches[2].id, moduleIds: [modules[4].id, modules[7].id, modules[1].id], status: 'IN_PROGRESS', completion: 45 },
    { batchId: batches[3].id, moduleIds: [modules[1].id, modules[2].id, modules[5].id], status: 'PENDING', completion: 10 },
  ];

  for (const bma of batchModuleAssignments) {
    for (const moduleId of bma.moduleIds) {
      await prisma.batchModule.upsert({
        where: { batchId_moduleId: { batchId: bma.batchId, moduleId } },
        update: {},
        create: {
          batchId: bma.batchId,
          moduleId,
          status: bma.status,
          completionPercent: bma.completion,
        },
      });
    }
  }
  console.log('Assigned modules to batches');

  // Trainer-batch assignments
  const trainerBatchAssignments = [
    { trainerId: trainers[0].id, batchId: batches[0].id },
    { trainerId: trainers[0].id, batchId: batches[1].id },
    { trainerId: trainers[1].id, batchId: batches[1].id },
    { trainerId: trainers[1].id, batchId: batches[3].id },
    { trainerId: trainers[2].id, batchId: batches[0].id },
    { trainerId: trainers[2].id, batchId: batches[2].id },
  ];

  for (const tba of trainerBatchAssignments) {
    await prisma.trainerBatch.upsert({
      where: { trainerId_batchId: { trainerId: tba.trainerId, batchId: tba.batchId } },
      update: {},
      create: {
        trainerId: tba.trainerId,
        batchId: tba.batchId,
      },
    });
  }
  console.log('Assigned trainers to batches');

  // Create 20 students
  const studentNames = [
    'Alice Johnson', 'Bob Smith', 'Charlie Brown', 'Diana Prince', 'Eve Davis',
    'Frank Miller', 'Grace Lee', 'Henry Wilson', 'Iris Chen', 'Jack Taylor',
    'Kate Anderson', 'Liam Thomas', 'Maya Patel', 'Noah Garcia', 'Olivia Martinez',
    'Peter Robinson', 'Quinn Walker', 'Rachel Hall', 'Sam Allen', 'Tara Young',
  ];

  const students = [];
  for (let i = 0; i < 20; i++) {
    const email = `student${i + 1}@institute.com`;
    const batchIndex = i % 4;
    const counsellorIndex = i % 2;
    const mode = i % 3 === 0 ? 'ONLINE' : 'OFFLINE';

    const studentUser = await prisma.user.upsert({
      where: { email },
      update: {},
      create: {
        email,
        password: studentPassword,
        name: studentNames[i],
        role: 'STUDENT',
        phone: `90000${(10000 + i + 1).toString()}`,
      },
    });

    const student = await prisma.student.upsert({
      where: { userId: studentUser.id },
      update: {},
      create: {
        userId: studentUser.id,
        batchId: batches[batchIndex].id,
        counsellorId: counsellors[counsellorIndex].id,
        mode,
      },
    });
    students.push(student);
  }
  console.log('Created 20 students');

  // Counsellor-student assignments
  for (let i = 0; i < students.length; i++) {
    const counsellorIndex = i % 2;
    await prisma.counsellorStudent.upsert({
      where: {
        counsellorId_studentId: {
          counsellorId: counsellors[counsellorIndex].id,
          studentId: students[i].id,
        },
      },
      update: {},
      create: {
        counsellorId: counsellors[counsellorIndex].id,
        studentId: students[i].id,
      },
    });
  }
  console.log('Assigned counsellors to students');

  // Create sample lectures
  const lectures = [];
  for (let week = 0; week < 4; week++) {
    for (let day = 0; day < 5; day++) {
      const date = new Date('2024-03-01');
      date.setDate(date.getDate() + week * 7 + day);

      const lecture = await prisma.lecture.create({
        data: {
          batchId: batches[0].id,
          moduleId: modules[week % modules.length].id,
          trainerId: trainers[0].id,
          date,
          startTime: '09:00',
          endTime: '12:00',
          duration: 180,
          topics: `Week ${week + 1}, Day ${day + 1}: ${modules[week % modules.length].name} session`,
        },
      });
      lectures.push(lecture);
    }
  }
  console.log(`Created ${lectures.length} lectures`);

  // Create attendance records for students in batch 1
  const batch1Students = students.filter((_, i) => i % 4 === 0);
  for (const lecture of lectures.slice(0, 10)) {
    for (const student of batch1Students) {
      const status = Math.random() > 0.2 ? 'PRESENT' : 'ABSENT';
      await prisma.attendance.create({
        data: {
          studentId: student.id,
          lectureId: lecture.id,
          status,
          method: 'MANUAL',
        },
      });
    }
  }
  console.log('Created attendance records');

  // Create marks for students
  const markTypes = ['QUIZ', 'ASSIGNMENT', 'PROJECT', 'EXAM'];
  for (const student of batch1Students) {
    for (let j = 0; j < 4; j++) {
      const mod = modules[j];
      await prisma.marks.create({
        data: {
          studentId: student.id,
          moduleId: mod.id,
          type: markTypes[j],
          score: Math.floor(Math.random() * 40) + 60,
          maxScore: 100,
          remarks: `${markTypes[j]} for ${mod.name}`,
        },
      });
    }
  }
  console.log('Created marks records');

  // Create fees for all students
  for (const student of students) {
    const totalAmount = 50000;
    const paidAmount = Math.floor(Math.random() * 4) * 12500;
    const pendingAmount = totalAmount - paidAmount;
    const status = paidAmount >= totalAmount ? 'PAID' : paidAmount > 0 ? 'PARTIAL' : 'PENDING';

    const fee = await prisma.fee.create({
      data: {
        studentId: student.id,
        totalAmount,
        paidAmount,
        pendingAmount,
        dueDate: new Date('2024-06-30'),
        status,
      },
    });

    // Create payment records
    if (paidAmount > 0) {
      const numPayments = paidAmount / 12500;
      for (let p = 0; p < numPayments; p++) {
        const paidAt = new Date('2024-01-15');
        paidAt.setMonth(paidAt.getMonth() + p);
        await prisma.feePayment.create({
          data: {
            feeId: fee.id,
            amount: 12500,
            paidAt,
            method: p % 2 === 0 ? 'UPI' : 'BANK_TRANSFER',
          },
        });
      }
    }
  }
  console.log('Created fee records');

  // Create notifications
  const notificationData = [
    { title: 'Welcome to the Institute', message: 'Your enrollment is confirmed. Please check your batch schedule.' },
    { title: 'New Module Started', message: 'React.js module is starting next week. Make sure to complete prerequisites.' },
    { title: 'Fee Reminder', message: 'Your next installment is due by end of this month.' },
    { title: 'Mock Interview Scheduled', message: 'Your mock interview has been scheduled for next Friday.' },
  ];

  for (const student of batch1Students) {
    const studentUser = await prisma.student.findUnique({
      where: { id: student.id },
      select: { userId: true },
    });
    if (studentUser) {
      for (const nd of notificationData) {
        await prisma.notification.create({
          data: {
            userId: studentUser.userId,
            title: nd.title,
            message: nd.message,
            type: 'INFO',
            isRead: Math.random() > 0.5,
          },
        });
      }
    }
  }
  console.log('Created notifications');

  // Create announcements
  await prisma.announcement.create({
    data: {
      createdById: admin.id,
      title: 'Holiday Notice',
      message: 'The institute will be closed on March 25th for annual function.',
      targetRole: 'STUDENT',
      batchId: batches[0].id,
    },
  });
  await prisma.announcement.create({
    data: {
      createdById: admin.id,
      title: 'New Batch Starting',
      message: 'A new batch for Data Science is starting on April 1st. Spread the word!',
      targetRole: 'TRAINER',
    },
  });
  console.log('Created announcements');

  console.log('\nSeeding completed successfully!');
  console.log('---');
  console.log('Default credentials:');
  console.log('  Admin: admin@institute.com / admin123');
  console.log('  Trainer: trainer1@institute.com / trainer123');
  console.log('  Counsellor: counsellor1@institute.com / counsellor123');
  console.log('  Student: student1@institute.com / student123');
}

main()
  .catch((e) => {
    console.error(e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
